compiler/rustc_resolve/src/late/lifetimes.rs

```diff
@@ -1270,97 +1231,117 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
+                            this.check_lifetime_params(old_scope, &bound_generic_params);
                             this.visit_ty(&bounded_ty);
                             walk_list!(this, visit_param_bound, bounds);
-                        }
-                    })
-                }
-                &hir::WherePredicate::RegionPredicate(hir::WhereRegionPredicate {
-                    ref lifetime,
-                    bounds,
-                    ..
-                }) => {
-                    self.visit_lifetime(lifetime);
-                    walk_list!(self, visit_param_bound, bounds);
-                }
-                &hir::WherePredicate::EqPredicate(hir::WhereEqPredicate {
-                    ref lhs_ty,
-                    ref rhs_ty,
-                    ..
-                }) => {
-                    self.visit_ty(lhs_ty);
-                    self.visit_ty(rhs_ty);
+                        })
+                    }
+                    &hir::WherePredicate::RegionPredicate(hir::WhereRegionPredicate {
+                        ref lifetime,
+                        bounds,
+                        ..
+                    }) => {
+                        this.visit_lifetime(lifetime);
+                        walk_list!(this, visit_param_bound, bounds);
+                    }
+                    &hir::WherePredicate::EqPredicate(hir::WhereEqPredicate {
+                        ref lhs_ty,
+                        ref rhs_ty,
+                        ..
+                    }) => {
+                        this.visit_ty(lhs_ty);
+                        this.visit_ty(rhs_ty);
+                    }
                 }
             }
-        }
+        })
     }
 
     fn visit_param_bound(&mut self, bound: &'tcx hir::GenericBound<'tcx>) {
+        // FIXME(jackh726): This is pretty weird. `LangItemTrait` doesn't go
+        // through the regular poly trait ref code, so we don't get another
+        // chance to introduce a binder. For now, I'm keeping the existing logic
+        // of "if there isn't a Binder scope above us, add one", but I
+        // imagine there's a better way to go about this.
+        let mut scope = self.scope;
+        let trait_ref_hack = loop {
+            match scope {
+                Scope::TraitRefBoundary { .. } | Scope::Body { .. } | Scope::Root => {
+                    break false;
+                }
+
+                Scope::Binder { .. } => {
+                    break true;
+                }
+
+                Scope::Elision { s, .. }
+                | Scope::ObjectLifetimeDefault { s, .. }
+                | Scope::Supertrait { s, .. } => {
+                    scope = s;
+                }
+            }
+        };
         match bound {
-            hir::GenericBound::LangItemTrait(_, _, hir_id, _) if self.trait_ref_hack.is_none() => {
+            hir::GenericBound::LangItemTrait(_, _, hir_id, _) if !trait_ref_hack => {
```

Comment:
I don't totally understand this logic. Why are we added an `Other` binder here? Why do we care if these is another binder in scope?

Reply:
I've changed this to be more in line with how it's done in `visit_poly_trat_ref`. I really don't know what the expected behavior here is though. And the fact that either way doesn't change things makes me wonder how much can be "removed" here. I would like it for this to just use the `visit_poly_trait_ref` path though.

Reply:
It seems to me that `LangItemTrait` is *meant* to be equivalent to the ordinary 'poly trait ref' path.